modules/kameleonvolume/kameleonvolumereader.cpp

```
@@ -0,0 +1,188 @@
+        return;
+    }
+
+    _model = _kameleon.model;
+    _interpolator = std::unique_ptr<ccmc::Interpolator>(_model->createNewInterpolator());
+}
+
+std::unique_ptr<RawVolume<float>> KameleonVolumeReader::readFloatVolume(
+    const glm::uvec3 & dimensions,
+    const std::string & variable,
+    const glm::vec3 & lowerBound,
+    const glm::vec3 & upperBound) const
+{
+    auto volume = std::make_unique<RawVolume<float>>(dimensions);
+    glm::vec3 dims = volume->dimensions();
+    glm::vec3 diff = upperBound - lowerBound;
+
+    _model->loadVariable(variable);
+
+    float* data = volume->data();
+    for (size_t index = 0; index < volume->nCells(); index++) {
+        glm::vec3 coords = volume->indexToCoords(index);
+        glm::vec3 coordsZeroToOne = coords / dims;
+        glm::vec3 volumeCoords = lowerBound + diff * coordsZeroToOne;
+
+        data[index] = _interpolator->interpolate(
+            variable,
+            static_cast<float>(volumeCoords[0]),
+            static_cast<float>(volumeCoords[1]),
+            static_cast<float>(volumeCoords[2]));
+    }
+
+    return volume;
+}
+
+std::vector<std::string> KameleonVolumeReader::gridVariableNames() const {
+    // get the grid system string
+    std::string gridSystem = _model->getGlobalAttribute("grid_system_1").getAttributeString();
+
+    // remove leading and trailing brackets
+    gridSystem = gridSystem.substr(1, gridSystem.length() - 2);
+
+    // remove all whitespaces
+    gridSystem.erase(remove_if(gridSystem.begin(), gridSystem.end(), isspace), gridSystem.end());
+
+    // replace all comma signs with whitespaces
+    std::replace(gridSystem.begin(), gridSystem.end(), ',', ' ');
+
+    // tokenize
+    std::istringstream iss(gridSystem);
+    std::vector<std::string> tokens{ std::istream_iterator<std::string>{iss},std::istream_iterator<std::string>{} };
+
+    // validate
+    if (tokens.size() != 3) {
+        return std::vector<std::string>();
```

Better to throw an exception here?